README.rst

```diff
@@ -1962,6 +1968,344 @@ However, if specifying :code:`Table` objects manually, please make sure to inclu
 for each table and to use only SQL types allowed for the dialect specified in the
 :code:`SQLAlchemySchemaInfo`.
 
+Macro system
+------------
+
+In many real-life situations, the database schema does not fit the user's mental model of the data.
+There are many causes of this, the most common one being database normalization.
+The representation of the data that is convenient for storage within a database is rarely
+the representation that makes for easy querying. As a result, users' queries frequently
+include complex and repetitive query structures that work around the database's chosen data model.
+
+The compiler's macro system empowers users *reshaping* their data's structure to fit
+their mental model, minimizing query complexity and repetitiveness without requiring changes
+to the shape of the data in the underlying data systems. The compiler achieves this by allowing
+users to define **macros** -- type-safe rules for programmatic query rewriting
+that transform user-provided queries on the *desired* data model into
+queries on the *actual* data model in the underlying data systems.
+
+When macros are defined, the compiler loads them into a macro registry -- a data structure used to
+track all currently available macros, the resulting GraphQL schema (accounting for macros), and
+any additional metadata needed by the compiler. The compiler then leverages this registry
+to expand queries that rely on macros, rewriting them into equivalent queries that do not contain
+any macros and therefore reflect the actual underlying data model. This makes them somewhat similar
+to SQL's idea of non-materialized views, though there are some key differences:
+
+- SQL views require database access and special permissions; databases are
+  completely oblivious to the use of macros since by the time the database gets the query,
+  all macro uses have been already expanded.
+
+- Macros can be stored and expanded client-side, so different users that query the same system may
+  define their own personal macros which are not shared with other users or the server that executes
+  the users' GraphQL queries. This is generally not achievable with SQL.
+
+- Since macro expansion does not interact in any way with the underlying data system, it works
+  seamlessly with all databases and even on schemas stitched together from multiple databases.
+  In contrast, not all databases support SQL-like :code:`VIEW` functionality.
+
+For now, the compiler supports one type of macro, **macro edges**. We hope to add more types
+of macros in the future.
+
+Macro registry
```

Comment:
I would start by giving an example of a macro edge as soon as possible. Like in https://github.com/kensho-technologies/graphql-compiler/pull/293/files

I would not talk about pro-forma fields in that example. I would probably have a small section to talk about the peculiarity of using macros with pro-forma fields.

Reply:
I agree with pmantica here on an early example. Perhaps you can add a sentence in the first paragraph before, "As a result," that says, "Imagine if you were.. xyz" or "Imagine trying to .. abc"

Reply:
resolved